File: 0698-partition-to-k-equal-sum-subsets/0698-partition-to-k-equal-sum-subsets.py
class Solution:
    def canPartitionKSubsets(self, arr: List[int], k: int) -> bool:
        n = len(arr)
        dp = [ -1 for i in range(1<< 16 + 2)]
        dp[0] = 0

        sum_ = sum(arr)
        if sum_ % k:
            return False
        tar = sum_ // k 

        for mask in range(1 << n):
            if dp[mask] == -1:
                continue
            for i in range(n):
                if (not(mask & 1 << i) and dp[mask] + arr[i] <= tar):
                    dp[mask | (1 << i)] = (dp[mask] + arr[i]) % tar
        
        return dp[(1 << n) -1 ] == 0

File: 0698-partition-to-k-equal-sum-subsets/0698-partition-to-k-equal-sum-subsets.cpp
class Solution {
public:
    bool canPartitionKSubsets(vector<int>& nums, int k) {
        int total_array_sum = 0;
        for (int num : nums){
            total_array_sum += num;
        }
        int n = nums.size();

        if (total_array_sum % k != 0){
            return false;
        }

        int target_sum = total_array_sum / k;

        vector<int> subsetSum((1 << n), -1);
        subsetSum[0] = 0;

        for (int mask = 0; mask < (1 << n); mask++){
            if (subsetSum[mask] == -1){
                continue;
            }

            for (int i = 0; i < n; i++){
                if ((mask & ( 1 << i)) == 0 & subsetSum[mask] + nums[i] <= target_sum){
                    subsetSum[mask | (1 << i)] = (subsetSum[mask] + nums[i]) % target_sum;
                }
            }

            if(subsetSum[(1 << n) - 1] == 0) { 
                return true;
            }
        }
        return subsetSum[(1 << n) - 1] == 0;
    }
};